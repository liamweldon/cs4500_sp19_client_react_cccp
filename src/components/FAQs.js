import React from 'react';
import FAQService from '../services/FAQService';
import './table.scss';
import {Link} from 'react-router-dom';
import './FAQs.scss';
import {omit} from 'lodash';

class FAQs extends React.Component {
  constructor(props) {
    super(props);
    this.faqService = FAQService.getInstance();
    this.state = {
      faqs: [],
      // key value pairs where the key is the ID and the value is the fields
      // ex {0: {title: '', question: ''}, 1: {title: '', question: ''}}
      editing: {},
      newQuestion: {title: '', question: ''}
    };
  }

  getFAQs = () => {
    this.faqService.findAllFAQs().then((faqs) =>
      this.setState({
        faqs: faqs
      })
    );
  };

  componentDidMount() {
    this.getFAQs();
  }

  deleteQuestion = (id) => {
    // once the question is deleted, reload the questions so the user doesnt see the deleted one
    this.faqService.deleteFAQ(id).then((res) => this.getFAQs());
  };

  saveQuestion = (id) => {
    const newQuestion = this.state.editing[id];
    this.faqService.editFAQ(id, newQuestion).then((res) => {
      this.getFAQs();
      this.setState({editing: omit(this.state.editing, id)});
    });
  };

  handleNewQuestionInputChange = (e) => {
    this.setState({newQuestion: {...this.state.newQuestion, [e.target.name]: e.target.value}});
  };

  addQuestion = () => {
    const newQuestion = this.state.newQuestion;
    this.faqService.addFAQ(newQuestion).then((res) => {
      this.getFAQs();
      this.setState({newQuestion: {title: '', question: ''}});
    });
  };

  handleInputChange = (e, id) => {
    let editingCopy = {...this.state.editing};
    editingCopy[id] = {...editingCopy[id], [e.target.name]: e.target.value};
    this.setState({editing: editingCopy});
  };

  toggleEditMode = (question) => {
    if (question.id in this.state.editing) {
      this.setState({editing: omit(this.state.editing, question.id)});
    } else {
      this.setState({
        editing: {...this.state.editing, [question.id]: {title: question.title, question: question.question}}
      });
    }
  };

  render() {
    return (
      <div>
        <h3>Frequently Asked Questions</h3>
        <table className="table faqs">
          <tbody>
            <tr className="header-row">
              <td>Title</td>
              <td>Question</td>
              <td />
              <td />
            </tr>
            <tr key={-1}>
              <td>
                <input value={this.state.newQuestion.title} name="title" onChange={this.handleNewQuestionInputChange} />
              </td>
              <td>
                <input
                  value={this.state.newQuestion.question}
                  name="question"
                  onChange={this.handleNewQuestionInputChange}
                />
              </td>
              <td>
                <button onClick={this.addQuestion}>
                  <i className="fas fa-plus-square" />
                </button>
              </td>
              <td />
            </tr>
            {this.state.faqs.map((faq) => (
              <tr key={faq.id}>
                <td>
                  {faq.id in this.state.editing ? (
                    <input
                      value={this.state.editing[faq.id].title}
                      name="title"
                      onChange={(e) => {
                        this.handleInputChange(e, faq.id);
                      }}
                    />
                  ) : (
                    <Link to={`/admin/faqs/${faq.id}`}>{faq.title}</Link>
                  )}
                </td>
                <td>
                  {faq.id in this.state.editing ? (
                    <input
                      defaultValue={this.state.editing[faq.id].question}
                      name="question"
                      onChange={(e) => {
                        this.handleInputChange(e, faq.id);
                      }}
                    />
                  ) : (
                    faq.question
                  )}
                </td>
                <td>
                  <button
                    onClick={() => {
                      this.deleteQuestion(faq.id);
                    }}
                  >
                    <i className="fas fa-trash-alt" />
                  </button>
                </td>
                <td>
                  {faq.id in this.state.editing ? (
                    <i
                      className="fas fa-check-square"
                      onClick={() => {
                        this.saveQuestion(faq.id);
                      }}
                    />
                  ) : (
                    <i className="fas fa-pen-square" onClick={() => this.toggleEditMode(faq)} />
                  )}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

export default FAQs;
