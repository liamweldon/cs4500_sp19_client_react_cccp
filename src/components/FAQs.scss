.faqs {
  button {
    background: none;
    border: none;
  }
  i {
    font-size: 2em;

    &.fa-trash-alt {
      color: red;
    }
    &.fa-pen-square,
    &.plus-square {
      color: blue;
    }
    &.check-square {
      color: green;
    }
  }
}
