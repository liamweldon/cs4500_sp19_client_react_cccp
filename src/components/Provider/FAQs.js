import React from 'react';
import './Provider.scss';

const FAQs = ({faqs}) => (
  <div className="faqs">
    {faqs.map((faq) => (
      <div className="faq">
        <h5>{faq.question}</h5>
        <div>{faq.answer}</div>
      </div>
    ))}
  </div>
);

export default FAQs;
