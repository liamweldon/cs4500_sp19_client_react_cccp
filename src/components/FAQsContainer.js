import React from 'react';
import FAQService from '../services/FAQService';
import './table.scss';
import './FAQs.scss';
import FAQs from './FAQs.js';
import {omit} from 'lodash';

class FAQsContainer extends React.Component {
  constructor(props) {
    super(props);
    this.faqService = FAQService.getInstance();
    this.state = {
      faqs: [],
      // key value pairs where the key is the ID and the value is the fields
      // ex {0: {title: '', question: ''}, 1: {title: '', question: ''}}
      editing: {},
      newQuestion: {title: '', question: ''}
    };
  }

  getFAQs = () => {
    this.faqService.findAllFAQs().then((faqs) => {
      if (faqs.length) {
        this.setState({
          faqs: faqs
        });
      }
    });
  };

  componentDidMount() {
    this.getFAQs();
  }

  deleteQuestion = (id) => {
    // once the question is deleted, reload the questions so the user doesnt see the deleted one
    this.faqService.deleteFAQ(id).then((res) => this.getFAQs());
  };

  saveQuestion = (id) => {
    const newQuestion = this.state.editing[id];
    this.faqService.editFAQ(id, newQuestion).then((res) => {
      this.getFAQs();
      this.setState({editing: omit(this.state.editing, id)});
    });
  };

  handleNewQuestionInputChange = (e) => {
    this.setState({newQuestion: {...this.state.newQuestion, [e.target.name]: e.target.value}});
  };

  addQuestion = () => {
    const newQuestion = this.state.newQuestion;
    this.faqService.addFAQ(newQuestion).then((res) => {
      this.getFAQs();
      this.setState({newQuestion: {title: '', question: ''}});
    });
  };

  handleInputChange = (e, id) => {
    let editingCopy = {...this.state.editing};
    editingCopy[id] = {...editingCopy[id], [e.target.name]: e.target.value};
    this.setState({editing: editingCopy});
  };

  toggleEditMode = (question) => {
    if (question.id in this.state.editing) {
      this.setState({editing: omit(this.state.editing, question.id)});
    } else {
      this.setState({
        editing: {...this.state.editing, [question.id]: {title: question.title, question: question.question}}
      });
    }
  };

  render() {
    const eventHandlers = {
      handleDeleteClick: this.deleteQuestion,
      handleNewQuestionInputChange: this.handleNewQuestionInputChange,
      handleInputChange: this.handleInputChange,
      handleSaveClick: this.saveQuestion,
      handleEditClick: this.toggleEditMode,
      handleAddClick: this.addQuestion
    };

    return (
      <FAQs
        faqs={this.state.faqs}
        editing={this.state.editing}
        newQuestion={this.state.newQuestion}
        eventHandlers={eventHandlers}
      />
    );
  }
}

export default FAQsContainer;
