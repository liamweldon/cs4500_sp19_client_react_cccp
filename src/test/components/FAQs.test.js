import React from 'react';
import FAQs from '../../components/FAQs';
import FAQsContainer from '../../containers/FAQsContainer';
import renderer from 'react-test-renderer';
import faqsJson from '../mockData/faq.mock.json';
import {StaticRouter} from 'react-router';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

test('FAQs Summary Snapshots', () => {
  const component = renderer.create(
    <StaticRouter location="/admin/faqs" context={{}}>
      <FAQs faqs={faqsJson} editing={{}} newQuestion={{title: '', question: ''}} eventHandlers={{}} />
    </StaticRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FAQs summary DOM', () => {
  let wrapper = mount(
    <StaticRouter location="/admin/faqs" context={{}}>
      <FAQsContainer />
    </StaticRouter>
  );

  let faqsComponent = wrapper.find(FAQsContainer);

  faqsComponent.getFAQs = jest.fn().mockReturnValue(faqsJson);
  wrapper.update();
  faqsComponent.setState({faqs: faqsJson});

  const firstFaqEditBtn = wrapper.find('.fa-pen-square').at(0);
  // there should be a row in the table for each FAQ
  expect(wrapper.find('.faq-row.existing').length).toEqual(faqsJson.length);
  expect(
    wrapper
      .find('.faq-row.existing')
      .at(0)
      .find('input').length
  ).toEqual(0);

  // click the pen icon to trigger edit mode for the first faq
  firstFaqEditBtn.props().onClick();
  wrapper.update();

  expect(
    wrapper
      .find('.faq-row.existing')
      .at(0)
      .find('input').length
  ).toEqual(2);
});
