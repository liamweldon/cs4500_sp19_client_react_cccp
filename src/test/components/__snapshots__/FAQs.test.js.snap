// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAQs Summary Snapshots 1`] = `
<div>
  <h3>
    Frequently Asked Questions
  </h3>
  <table
    className="table faqs"
  >
    <tbody>
      <tr
        className="header-row"
      >
        <td>
          Title
        </td>
        <td>
          Question
        </td>
        <td />
        <td />
      </tr>
      <tr
        className="faq-row new"
      >
        <td>
          <input
            name="title"
            value=""
          />
        </td>
        <td>
          <input
            name="question"
            value=""
          />
        </td>
        <td>
          <button>
            <i
              className="fas fa-plus-square"
            />
          </button>
        </td>
        <td />
      </tr>
      <tr
        className="faq-row existing"
      >
        <td>
          <a
            href="/admin/faqs/172"
            onClick={[Function]}
          >
            Background Check
          </a>
        </td>
        <td>
          Have you completed a background check?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        className="faq-row existing"
      >
        <td>
          <a
            href="/admin/faqs/192"
            onClick={[Function]}
          >
            Employees
          </a>
        </td>
        <td>
          How many employees does your company have?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        className="faq-row existing"
      >
        <td>
          <a
            href="/admin/faqs/232"
            onClick={[Function]}
          >
            Games
          </a>
        </td>
        <td>
          You got games on your phone?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`FAQs summary DOM 1`] = `
<div>
  <h3>
    Frequently Asked Questions
  </h3>
  <table
    className="table faqs"
  >
    <tbody>
      <tr
        className="header-row"
      >
        <td>
          Title
        </td>
        <td>
          Question
        </td>
        <td />
        <td />
      </tr>
      <tr
        className="faq-row new"
      >
        <td>
          <input
            name="title"
            value=""
          />
        </td>
        <td>
          <input
            name="question"
            value=""
          />
        </td>
        <td>
          <button>
            <i
              className="fas fa-plus-square"
            />
          </button>
        </td>
        <td />
      </tr>
      <tr
        className="faq-row existing"
      >
        <td>
          <a
            href="/admin/faqs/172"
            onClick={[Function]}
          >
            Background Check
          </a>
        </td>
        <td>
          Have you completed a background check?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        className="faq-row existing"
      >
        <td>
          <a
            href="/admin/faqs/192"
            onClick={[Function]}
          >
            Employees
          </a>
        </td>
        <td>
          How many employees does your company have?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        className="faq-row existing"
      >
        <td>
          <a
            href="/admin/faqs/232"
            onClick={[Function]}
          >
            Games
          </a>
        </td>
        <td>
          You got games on your phone?
        </td>
        <td>
          <button
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt"
            />
          </button>
        </td>
        <td>
          <i
            className="fas fa-pen-square"
            onClick={[Function]}
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
